src/source.ts

```diff
@@ -0,0 +1,66 @@
+import * as React from 'react';
+import {
+  Map,
+  VectorSource,
+  RasterSource,
+  GeoJSONSource,
+  ImageSource,
+  VideoSource,
+  GeoJSONSourceRaw
+} from 'mapbox-gl/dist/mapbox-gl';
+
+export interface Context {
+  map: Map;
+}
+
+export type Sources = VectorSource | RasterSource | GeoJSONSource | ImageSource | VideoSource | GeoJSONSourceRaw;
+
+export interface Props {
+  id: string;
+  sourceOptions: Sources;
+}
+
+export default class Source extends React.Component<Props, void> {
+  public context: Context;
+
+  public static contextTypes = {
+    map: React.PropTypes.object
+  };
+
+  private id = this.props.id;
+
+  public componentWillMount() {
+    const { map } = this.context;
+    if (!map.getSource(this.id)) {
+      map.addSource(this.id, this.props.sourceOptions);
+    }
+  }
+
+  public componentWillUnmount() {
+    const { map } = this.context;
+    if (map.getSource(this.id)) {
+      map.removeSource(this.id);
+    }
+  }
+
+  public componentWillReceiveProps(props: Props) {
+    const { id } = this;
+    const { sourceOptions } = this.props;
+    const { map } = this.context;
+
+    if ((props.sourceOptions as GeoJSONSourceRaw).data !== (sourceOptions as GeoJSONSourceRaw).data) {
+      (map
+        .getSource(id) as GeoJSONSource)
+        .setData((props.sourceOptions as GeoJSONSourceRaw).data as any);
+    }
```

Comment:
It's a bit tricky here.
@alex3165 
First, simplify sourceOptions type. You don't really need this : 

`export type Sources = VectorSource | RasterSource | GeoJSONSource | ImageSource | VideoSource | GeoJSONSourceRaw;`

Just do this : 
`sourceOptions: GeoJSONSourceRaw;`

And then you can write this line like this : 
```javascript
if ((props.sourceOptions).data !== (sourceOptions).data) {
      const source = map.getSource(id) as GeoJSONSource;
      const data = props.sourceOptions.data as GeoJSON.Feature<GeoJSON.GeometryObject> | GeoJSON.FeatureCollection<GeoJSON.GeometryObject> | string;
      source.setData(data);
}
```

By the way the long type `GeoJSON.Feature<GeoJSON.GeometryObject> | GeoJSON.FeatureCollection<GeoJSON.GeometryObject> | string` is also present in `geojson-layer.ts`, so you can make a type alias of this.